src/Common/Thermodynamics/relations.jl

```diff
@@ -1353,6 +1378,96 @@ function saturation_adjustment(
     end
 end
 
+"""
+    saturation_adjustment_ρpq(
+        param_set,
+        ρ,
+        p,
+        q_tot,
+        phase_type,
+        maxiter,
+        temperature_tol
+    )
+Compute the temperature that is consistent with
+ - `param_set` an `AbstractParameterSet`, see the [`Thermodynamics`](@ref) for more details
+ - `ρ` (moist-)air density
+ - `p` pressure
+ - `q_tot` total specific humidity
+ - `phase_type` a thermodynamic state type
+ - `maxiter` maximum iterations for non-linear equation solve
+ - `temperature_tol` temperature tolerance
+by finding the root of
+
+```
+T - air_temperature_from_ideal_gas_law(
+        param_set,
+        p,
+        ρ,
+        PhasePartition_equil(param_set, T, ρ, q_tot, phase_type),
+    )
+```
+using Newtons method using ForwardDiff.
+See also [`saturation_adjustment`](@ref).
+"""
+function saturation_adjustment_ρpq(
+    param_set::APS,
+    ρ::FT,
+    p::FT,
+    q_tot::FT,
+    phase_type::Type{<:PhaseEquil},
+    maxiter::Int,
+    temperature_tol::FT = eps(FT),
+) where {FT <: Real}
+    tol = SolutionTolerance(temperature_tol)
+    q_pt = PhasePartition(q_tot)
+    T_0 = air_temperature_from_ideal_gas_law(param_set, p, ρ, q_pt)
+    # Use `oftype` to preserve diagonalized type signatures:
+    sol = find_zero(
+        T ->
+            T - air_temperature_from_ideal_gas_law(
+                param_set,
+                oftype(T, p),
+                oftype(T, ρ),
+                PhasePartition_equil(
+                    param_set,
+                    T,
+                    oftype(T, ρ),
+                    oftype(T, q_tot),
+                    phase_type,
+                ),
+            ),
+        NewtonsMethodAD(T_0),
```

Comment:
If we actually want to use it, I suspect the secant method would be considerably faster for plausible tolerances. Did you test this?

Reply:
I wasn't sure what bounds to use for SecantMethod. `NewtonsMethodAD` converges in 5 iterations for all thermo profiles, so I figured that is reasonable.

Reply:
Ok. We'll not actually use it, I don't think. But the AD itself will also add cost per iteration. 

For the secant method, you could use the same kinds of bounds we had in other places for SA (use the unsaturated solution as one bound).

Reply:
Did you try Newton AD vs Newton with analytical gradients for the standard SA? I'm wondering about the timing. If AD is fast enough, it would be nice because it would allow us to use arbitrary (smooth) phase partitioning functions.

Reply:
> Ok. We'll not actually use it, I don't think. But the AD itself will also add cost per iteration.

Agreed, I figured that this will be a nice test to have, and provide transparency for the approximation that we're making.

> For the secant method, you could use the same kinds of bounds we had in other places for SA (use the unsaturated solution as one bound).

I can give that a try.

> Did you try Newton AD vs Newton with analytical gradients for the standard SA? I'm wondering about the timing.

I haven't tried yet, but that's a great point! Maybe that'll be better for a separate PR? I can open an issue to track it.

> If AD is fast enough, it would be nice because it would allow us to use arbitrary (smooth) phase partitioning functions.

This would be cool to try out also!

Reply:
Opened #1831.

Reply:
Using `BenchmarkTools`'s `@btime` on the thermo profiles, I'm seeing
```julia
[Float32, Newton AD]  1.206 ms (26298 allocations: 513.67 KiB)
[Float32, Secant   ]  1.429 ms (39854 allocations: 784.30 KiB)
[Float64, Newton AD]  1.492 ms (31110 allocations: 957.42 KiB)
[Float64, Secant   ]  1.894 ms (47694 allocations: 1.01 MiB)
```
So, it looks like Newton method with AD is performing better, which makes sense because Secant method seems to be requiring > 10 iterations to converge on the entire set of thermo profiles. I'll leave Newton method AD as the default for now.